src/libNetwork/ReputationManager.cpp

```diff
@@ -0,0 +1,164 @@
+/**
+* Copyright (c) 2018 Zilliqa
+* This source code is being disclosed to you solely for the purpose of your participation in
+* testing Zilliqa. You may view, compile and run the code for that purpose and pursuant to
+* the protocols and algorithms that are programmed into, and intended by, the code. You may
+* not do anything else with the code without express permission from Zilliqa Research Pte. Ltd.,
+* including modifying or publishing the code (or any part of it), and developing or forming
+* another public or private blockchain network. This source code is provided ‘as is’ and no
+* warranties are given as to title or non-infringement, merchantability or fitness for purpose
+* and, to the extent permitted by law, all liability for your use of the code is disclaimed.
+* Some programs in this code are governed by the GNU General Public License v3.0 (available at
+* https://www.gnu.org/licenses/gpl-3.0.en.html) (‘GPLv3’). The programs that are governed by
+* GPLv3.0 are those programs that are located in the folders src/depends and tests/depends
+* and which include a reference to GPLv3 in their program files.
+**/
+
+#include "ReputationManager.h"
+
+#include "Blacklist.h"
+#include "libUtils/IPConverter.h"
+#include "libUtils/Logger.h"
+#include "libUtils/SafeMath.h"
+
+ReputationManager::ReputationManager() {}
+
+ReputationManager::~ReputationManager() {}
+
+ReputationManager& ReputationManager::GetInstance()
+{
+    static ReputationManager RM;
+    return RM;
+}
+
+bool ReputationManager::IsNodeBanned(
+    const boost::multiprecision::uint128_t& IPAddress)
+{
+    if (GetReputation(IPAddress) <= REPTHRESHOLD)
+    {
+        return true;
+    }
+    return false;
+}
+
+void ReputationManager::PunishNode(
+    const boost::multiprecision::uint128_t& IPAddress, int32_t Penalty)
+{
+    UpdateReputation(IPAddress, Penalty);
+    if (!Blacklist::GetInstance().Exist(IPAddress) and IsNodeBanned(IPAddress))
+    {
+        LOG_GENERAL(INFO,
+                    "Node " << IPConverter::ToStrFromNumericalIP(IPAddress)
+                            << " banned.");
+        Blacklist::GetInstance().Add(IPAddress);
+    }
+}
+
+void ReputationManager::AwardAllNodes()
+{
+    std::vector<boost::multiprecision::uint128_t> AllKnownIPs = GetAllKnownIP();
+    for (const auto& ip : AllKnownIPs)
+    {
+        AwardNode(ip);
+    }
+}
+
+void ReputationManager::AddNodeIfNotKnown(
+    const boost::multiprecision::uint128_t& IPAddress)
+{
+    std::lock_guard<std::mutex> lock(m_mutexReputations);
```

Comment:
Might be worth having two versions of this method. One public that acquires the `m_mutexReputations` and one used privately that doesn't. Then you can use it in other methods like `GetReputation`, `SetReputation` *after* locking this avoid locking twice.